### PR TITLE
Port reverse

### LINKED ISSUE #1247: Port reverse
```c
DEFUN ("reverse", Freverse, Sreverse, 1, 1, 0,
       doc: /* Return the reversed copy of list, vector, or string SEQ.
See also the function `nreverse', which is used more often.  */)
  (Lisp_Object seq)
{
  Lisp_Object new;

  if (NILP (seq))
    return Qnil;
  else if (CONSP (seq))
    {
      new = Qnil;
      FOR_EACH_TAIL (seq)
	new = Fcons (XCAR (seq), new);
      CHECK_LIST_END (seq, seq);
    }
  else if (VECTORP (seq))
    {
      ptrdiff_t i, size = ASIZE (seq);

      new = make_uninit_vector (size);
      for (i = 0; i < size; i++)
	ASET (new, i, AREF (seq, size - i - 1));
    }
  else if (BOOL_VECTOR_P (seq))
    {
      ptrdiff_t i;
      EMACS_INT nbits = bool_vector_size (seq);

      new = make_uninit_bool_vector (nbits);
      for (i = 0; i < nbits; i++)
	bool_vector_set (new, i, bool_vector_bitref (seq, nbits - i - 1));
    }
  else if (STRINGP (seq))
    {
      ptrdiff_t size = SCHARS (seq), bytes = SBYTES (seq);

      if (size == bytes)
	{
	  ptrdiff_t i;

	  new = make_uninit_string (size);
	  for (i = 0; i < size; i++)
	    SSET (new, i, SREF (seq, size - i - 1));
	}
      else
	{
	  unsigned char *p, *q;

	  new = make_uninit_multibyte_string (size, bytes);
	  p = SDATA (seq), q = SDATA (new) + bytes;
	  while (q > SDATA (new))
	    {
	      int ch, len;

	      ch = STRING_CHAR_AND_LENGTH (p, len);
	      p += len, q -= len;
	      CHAR_STRING (ch, q);
	    }
	}
    }
  else
    wrong_type_argument (Qsequencep, seq);
  return new;
}
```

### PR DESCRIPTION
**Pull request by agraven**
Closes #1247.

The code doing the actual reversing for each of the different types could be methods that this function calls. I'm not sure which is preferable.

### CODE REVIEW

File: rust_src/src/fns.rs [hunk 1]
```diff
@@ -274,6 +278,67 @@ pub fn concat(args: &mut [LispObject]) -> LispObject {
     }
 }
 
+/// Return the reversed copy of list, vector, or string SEQ.
+/// See also the function `nreverse', which is used more often.
+#[lisp_fn]
+pub fn reverse(seq: LispObject) -> LispObject {
+    if seq.is_nil() {
+        return Qnil;
+    } else if let Some(cons) = seq.as_cons() {
```
Comment (shaleh) [inline rust_src/src/fns.rs]: This looks like a classic 'fold'. Also, I think `iter_cars` works here. `iter_tails` returns Cons cells and `iter_cars` is `XCAR(cons)` over `iter_tails`. Why the `into` at the end? `new` is already a LispObject right?

```
let new = cons.iter_cars(LispConsEndChecks::on, LispConsCircularChecks::on).fold(Qnil, |cons, elt| {
    LispObject::cons(elt, cons)
});
```

File: rust_src/src/multibyte.rs [hunk 2]
```diff
@@ -583,6 +583,30 @@ pub fn multibyte_char_at(slice: &[c_uchar]) -> (Codepoint, usize) {
     }
 }
 
+/// Same as STRING_CHAR_AND_LENGHT
```
Comment (shaleh) [inline rust_src/src/multibyte.rs]:
```suggestion
/// Same as STRING_CHAR_AND_LENGTH
```
Typo

File: rust_src/src/multibyte.rs [hunk 3]
```diff
@@ -583,6 +583,30 @@ pub fn multibyte_char_at(slice: &[c_uchar]) -> (Codepoint, usize) {
...
+        }
+        3 => {
+            let cp = (((head & 0x0F) as Codepoint) << 12)
+                | (((*ptr.add(1) & 0x3F) as Codepoint) << 6)
+                | (*ptr.add(2) & 0x3F) as Codepoint;
+            (cp, 3)
+        }
+        _ => {
+            let mut len = 0;
+            let cp = string_char(ptr, ptr::null_mut(), &mut len);
```
Comment (shaleh) [inline rust_src/src/multibyte.rs]: `string_char` should be migrated to a shallow C wrapper and a Rust function which returns the tuple. That can be another PR.

File: rust_src/src/fns.rs [hunk 4]
```diff
@@ -274,6 +278,67 @@ pub fn concat(args: &mut [LispObject]) -> LispObject {
     }
 }
 
+/// Return the reversed copy of list, vector, or string SEQ.
+/// See also the function `nreverse', which is used more often.
+#[lisp_fn]
+pub fn reverse(seq: LispObject) -> LispObject {
+    if seq.is_nil() {
+        return Qnil;
```
Comment (shaleh) [inline rust_src/src/fns.rs]:
```suggestion
        Qnil
```

You don't need the return since you made this all one if/else construct.

**Comment by shaleh** [comment]
re: methods or not.

If the methods existed, would the code benefit from calling them directly instead of calling `Freverse` or is this mostly called from lisp? If this is mostly lisp then the methods might be overkill for now. I could see them being beneficial if we could write Rust unit tests for the data types.

File: rust_src/src/fns.rs (hunk 1 above)
Comment (agraven) [reply rust_src/src/fns.rs]: Yeah, that looks like it would work. Would remove the need for a `new` as well.

File: rust_src/src/fns.rs (hunk 4 above)
Comment (agraven) [reply rust_src/src/fns.rs]: Fixed

File: rust_src/src/multibyte.rs (hunk 2 above)
Comment (agraven) [reply rust_src/src/multibyte.rs]: Thought something looked off about the way I wrote that. Thanks.

File: rust_src/src/fns.rs (hunk 1 above)
Comment (shaleh) [reply rust_src/src/fns.rs]: If we wanted to be cute we could even write it as `(elt, cons).into()`. I am ok with the explicit LispObject::cons().

Comment (agraven) [reply rust_src/src/fns.rs]: Yeah, I think the explicitness makes it easier to read.

**Comment by agraven** [comment]
The function is used very sparsely in C code (4 times), and not at all in rust code, so at least for now it should probably be left in the function, and can be moved out to methods if the need arises.

**Comment by shaleh** [comment]
`nreverse` is used far more. Is there a way to unify the code and support both? The difference is `nreverse` is destructive and `reverse` returns a new copy.

It looks like String and Cons are the two candidates for turning into methods once `nreverse` is also ported.

**Comment by agraven** [comment]
I suppose you could make reverse just call nreverse on a copy of its argument. I was actually just about to port nreverse since I was already at it, I'll look into the possibilities.

**Comment by shaleh** [comment]
Agreed. I wonder if an internal function which tasks `source` and `destination` where destination is sometimes the `make_uninit...` would work.
